File: nginx_test/was-7-flask/src/cache.py
import os
import json
import logging
from redis.cluster import RedisCluster

logger = logging.getLogger('was-7-flask')

REDIS_NODES = os.getenv('REDIS_NODES', 'redis-1:6379,redis-2:6379,redis-3:6379')
REDIS_PASSWORD = os.getenv('REDIS_PASSWORD', '')


class CacheService:
    def __init__(self):
        nodes = []
        for node in REDIS_NODES.split(','):
            host, port = node.split(':')
            nodes.append({'host': host, 'port': int(port)})

        try:
            self.client = RedisCluster(
                startup_nodes=nodes,
                decode_responses=True,
                password=REDIS_PASSWORD or None,
                skip_full_coverage_check=True
            )
            logger.info('Redis cluster connected')
        except Exception as e:
            logger.error(f'Failed to connect to Redis: {e}')
            self.client = None

    def get(self, key):
        if not self.client:
            return None
        try:
            value = self.client.get(key)
            logger.debug(f'Cache GET: key={key}, found={value is not None}')
            return value
        except Exception as e:
            logger.error(f'Cache GET error: key={key}, error={e}')
            return None

    def set(self, key, value, ttl=3600):
        if not self.client:
            raise Exception('Redis not connected')
        try:
            str_value = json.dumps(value) if isinstance(value, (dict, list)) else str(value)
            self.client.set(key, str_value, ex=ttl)
            logger.debug(f'Cache SET: key={key}, ttl={ttl}')
            return True
        except Exception as e:
            logger.error(f'Cache SET error: key={key}, error={e}')
            raise

    def delete(self, key):
        if not self.client:
            return False
        try:
            result = self.client.delete(key)
            logger.debug(f'Cache DELETE: key={key}, deleted={result > 0}')
            return result > 0
        except Exception as e:
            logger.error(f'Cache DELETE error: key={key}, error={e}')
            return False

    def health_check(self):
        if not self.client:
            return {'status': 'DOWN', 'error': 'Not connected'}
        try:
            result = self.client.ping()
            return {'status': 'UP' if result else 'DOWN', 'response': str(result)}
        except Exception as e:
            return {'status': 'DOWN', 'error': str(e)}

File: nginx_test/was-8-go/internal/cache/cache.go
package cache

import (
	"context"
	"os"
	"strings"
	"time"

	"github.com/redis/go-redis/v9"
	"go.uber.org/zap"
)

type CacheService struct {
	client *redis.ClusterClient
	logger *zap.SugaredLogger
}

func NewCacheService() *CacheService {
	zapLogger, _ := zap.NewProduction()
	sugar := zapLogger.Sugar()

	nodes := os.Getenv("REDIS_NODES")
	if nodes == "" {
		nodes = "redis-1:6379,redis-2:6379,redis-3:6379"
	}

	addrs := strings.Split(nodes, ",")

	client := redis.NewClusterClient(&redis.ClusterOptions{
		Addrs:        addrs,
		Password:     os.Getenv("REDIS_PASSWORD"),
		DialTimeout:  5 * time.Second,
		ReadTimeout:  3 * time.Second,
		WriteTimeout: 3 * time.Second,
		PoolSize:     50,
		MinIdleConns: 10,
	})

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	if err := client.Ping(ctx).Err(); err != nil {
		sugar.Warnf("Failed to connect to Redis cluster: %v", err)
	} else {
		sugar.Info("Connected to Redis cluster")
	}

	return &CacheService{
		client: client,
		logger: sugar,
	}
}

func (s *CacheService) Get(key string) (string, bool) {
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()

	value, err := s.client.Get(ctx, key).Result()
	if err == redis.Nil {
		return "", false
	}
	if err != nil {
		s.logger.Errorf("Cache GET error: key=%s, error=%v", key, err)
		return "", false
	}

	s.logger.Debugf("Cache GET: key=%s, found=true", key)
	return value, true
}

func (s *CacheService) Set(key, value string, ttl time.Duration) error {
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()

	err := s.client.Set(ctx, key, value, ttl).Err()
	if err != nil {
		s.logger.Errorf("Cache SET error: key=%s, error=%v", key, err)
		return err
	}

	s.logger.Debugf("Cache SET: key=%s, ttl=%v", key, ttl)
	return nil
}

func (s *CacheService) Delete(key string) bool {
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()

	result, err := s.client.Del(ctx, key).Result()
	if err != nil {
		s.logger.Errorf("Cache DELETE error: key=%s, error=%v", key, err)
		return false
	}

	s.logger.Debugf("Cache DELETE: key=%s, deleted=%v", key, result > 0)
	return result > 0
}

func (s *CacheService) HealthCheck() map[string]interface{} {
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()

	result, err := s.client.Ping(ctx).Result()
	if err != nil {
		return map[string]interface{}{
			"status": "DOWN",
			"error":  err.Error(),
		}
	}

	return map[string]interface{}{
		"status":   "UP",
		"response": result,
	}
}

func (s *CacheService) Close() {
	if s.client != nil {
		s.client.Close()
	}
}
